feat: add `/pallets/staking/validators` endpoint

Imod7 opened the pull request:
Relates to this Issue #1005 

### Description
Adds `/pallets/staking/validators` endpoint which returns a list of all validators addresses and their corresponding status which can be either `active` or `waiting`. 
* `Active` means that the validator is part of the active set
* `Waiting` means that the validator did not get into the active set this era


## Sample Response

Sample request :
`/pallets/staking/validators`

Sample response :

```
{
  "validators": [
    {
      "address": "HYxzaXwkydDNAy3R8UnCvfprNm74TwwSELpyXSRoqmy9AQY",
      "status": "waiting"
    },
    {
      "address": "DfishveZoxSRNRb8FtyS7ignbw6cr32eCY2w6ctLDRM1NQz",
      "status": "active"
    },
    {
      "address": "HmsTAS1bCtZc9FSq9nqJzZCEkhhSygtXj9TDxNgEWTHnpyQ",
      "status": "active"
    },
    {
      "address": "Efishjj6mrTWVjwFoxtMmHH3mX6tRNnPKsSfQeCLZSv3jAE",
      "status": "waiting"
    },
    {
      "address": "Hn8ELsyUL9UgDZpPX6bE464o4sbi7gqgYEKjntK7xuVzGXH",
      "status": "active"
    },
...
```

## Todos
- [x] Update Docs
- [x] Add Tests

Imod7 committed:
Added the `validators` endpoint
- Controller, service, response type
- Returns all validators addresses and their corresponding status (active / waiting)

TarikGul commented on src/controllers/pallets/PalletsStakingValidatorsController.ts:
```diff
@@ -0,0 +1,52 @@
+	 * @param _req Express Request
+	 * @param res Express Response
+	 */
+	private getPalletStakingValidators: RequestHandler = async (
```
As chatted about, lets have this accept a query param of `at`

TarikGul commented on src/services/pallets/PalletsStakingValidatorsService.ts:
```diff
@@ -0,0 +1,58 @@
+		const { api } = this;
+
+		// Validators in the active set
+		const validatorSession = await api.query.session.validators();
```
This api should be a result of `api.at(blockHash)` and named `historicApi`

TarikGul commented:
closes: https://github.com/paritytech/substrate-api-sidecar/issues/301

Imod7 committed:
Validators Info per block
- Added the `at` query param
- Modified the Service accordingly

Imod7 committed:
yarn lint

jsdw commented on src/services/pallets/PalletsStakingValidatorsService.ts:
```diff
@@ -0,0 +1,63 @@
+		const historicApi = await api.at(hash);
+
+		// Validators in the active set
+		const validatorSession = await historicApi.query.session.validators();
+		let validatorSet: string[] = [];
+		validatorSet = validatorSession.map((address) => address.toString());
```
Let's make this an actual `Set`, so something like:

```ts
let validatorSet: Set<string> = new Set();
for (const addres of validatorSession) validatorSet.insert(address.toString());
```

And then below, `validatorSet.contains(address)` instead of `validatorSet.includes(address)`.

(Random note; I think we have ~1000 validators in total)

Have a think about why this is really beneficial :)

Imod7 committed:
Changes `array` to `Set` (James Super Tip)
- Changed the data structure used to save the validators in the active set from an `array` to a `Set` (based on James suggestion)
- Changed also the used functions accordingly.
- Merged in one loop the check of which validator is in the active set and the population of the returned array.

Imod7 committed:
Merge branch 'master' into domi-validators-endpoint

Imod7 committed:
Lint & Update docs
- yarn lint
- Added validators endpoint to the docs
- Corrected other parts of the docs mentioned by the swagger as errors.

TarikGul commented on src/services/pallets/PalletsStakingValidatorsService.ts:
```diff
@@ -0,0 +1,62 @@
+		const { api } = this;
+		const historicApi = await api.at(hash);
+
+		// Querying to get the validators from the active set and populating
```
I don't think we need these comments. The code here is self documenting.

TarikGul commented on src/services/pallets/PalletsStakingValidatorsService.ts:
```diff
@@ -0,0 +1,62 @@
+		const validators: IValidator[] = [];
+		const validatorsEntries =
+			await historicApi.query.staking.validators.entries();
+		validatorsEntries.forEach(([key]) => {
```
This can be rewritten as a `map`. Something like:

`const validators = validatorsEntries.map(([key]) => {...})`

TarikGul reviewed (changes requested):
Looks great so far, just some nits, and code changes.

Imod7 committed:
Added in latest e2e tests the validators endpoint

Imod7 committed:
Changes from Tarik s comments

Imod7 commented:
@TarikGul @IkerAlus If validators can change status (go from `active` to `waiting` for example) per `era` then I have to add as `param` in the endpoint also the `era` right ?
From [Polkadot Support](https://support.polkadot.network/support/solutions/articles/65000168050-what-is-an-era-) 
> _An era is a period of time during which there is a specific set of active validators. Each era has 6 epochs (or sessions). During the last epoch, the active set of the next era is elected._

Imod7 committed:
Adjustments in the docstring

IkerAlus commented:
@Imod7 Either you add `blockHeight` or `era` as input parameter but not both. There is an unequivocally relationship between a block and an era. 

If the question is what is more useful as input parameter, block heights or eras, I am not sure, I see reasons for both:
- block height is used everywhere in the endpoints and it is easy to find
- era are more related to this endpoint

TarikGul commented on src/services/pallets/PalletsStakingValidatorsService.ts:
```diff
@@ -0,0 +1,61 @@
+		const historicApi = await api.at(hash);
+
+		const validatorSession = await historicApi.query.session.validators();
+		const validatorsActiveSet: Set<string> = new Set();
```
```suggestion
		const validatorsActiveSet= new Set<string>();
```

TarikGul reviewed (approved):
Just one nit. Great Job! lgtm

marshacb reviewed (approved):
looks great to me! once merge conflicts and any remaining nits are addressed its good to go!

Imod7 committed:
Update src/services/pallets/PalletsStakingValidatorsService.ts

Co-authored-by: Tarik Gul <47201679+TarikGul@users.noreply.github.com>

Imod7 committed:
Merged master & resolved conflicts

Imod7 committed:
yarn lint:fix
